### DISCUSSION TITLE
URL.include_query_params overwrites query string parameters with the same name

### DISCUSSION THREAD

**Discussion by lexabug** [encode/starlette#1735] [Potential Issue]
I've noticed that the `URL.include_query_params()` overwrites query string parameters with the same name. For example:
```python
from starlette.datastructures import URL, MultiDict

url = URL('my_test_url_example')
query_params = MultiDict([('my_id', '143155'), ('language', 'en'), ('list_p[]', 'item1'), ('list_p[]', 'original_name'), ('list_p[]', 'item_3')])
new_url = url.include_query_params(**query_params)
```

So the `new_url` will be:
```
URL('my_url_example?my_id=143155&language=en&list_p%5B%5D=item_3')
```
Which is incorrect. The specification of the URLs allows multiple query string parameters with the same name, so the backend that should process that query string must correctly treat such parameters as arrays/lists.

**Comment by jhominal** [comment]
I think that this issue is due to the way that, as `MultiDict` is unpacked with `**`, that call basically makes it so that that call is equivalent to:

```
new_url = url.include_query_params(**dict(query_params.items()))
```

Basically, `MultiDict` implements the `Mapping` interface by restricting the `Mapping` interface to accessing a single item for each key. That is for ease of use (in the very common case where each query string key only has a single value), but it does fail in this kind of usage.

Do we want to add a provision for that kind of usage? E.g. there could be an overload:

```python
def include_query_params(self, *items: typing.Union[
            ImmutableMultiDict[str, str]",
            typing.Mapping[str, str],
            typing.Iterable[typing.Tuple[str, str]],
        ],
): ...
```

**Reply by lexabug** [reply]
That would be great!

**Comment by lexabug** [comment]
Here is the PR for that https://github.com/encode/starlette/pull/1738

**Comment by GPla** [comment]
The PR was closed and no alternative is added natively. 
Here is a quick function which uses the fix from [here](https://github.com/alex-oleshkevich/ohmyadmin/blob/master/ohmyadmin/ordering.py#L78).


```python
def include_query_params(
    url: URL,
    **kwargs,
) -> URL:
    """Works like `URL.include_query_params` but correctly encodes parameters
    of type `list` and `set`.

    Parameters:
        url (URL): Url to update.
        **kwargs: Keyword arguments representing the query parameters to be added or modified.

    Returns:
        URL: Returns the url with the updated query parameters.
    """

    params = MultiDict(parse_qsl(url.query, keep_blank_values=True))
    for name, value in kwargs.items():
        if isinstance(value, list | set):
            params.setlist(name, list(value))
        else:
            params[name] = value
    url = url.replace(query=urlencode(params.multi_items()))
    return url
```

Use as:

```python
url = URL("http://localhost:8000/files")
include_query_params(url, param=[1, 2, 3], q="search")
# URL('http://localhost:8000/files?param=1&param=2&param=3&q=search')
```

Also easily usable inside templates, just add it to the globals of the template engine, e.g., `Jinja2`:
```python
templates.env.globals.update({'include_query_params': include_query_params})
```
